src/libvirtApi/domain.ts

```diff
@@ -115,6 +115,234 @@ function buildConsoleVVFile(consoleDetail: VMGraphics): string {
+    return cockpit.spawn(
+        args,
+        {
+            err: "message",
+            ...(connectionName === "system" ? { superuser: "try" } : { })
+        });
+}
+
+function script(connectionName: ConnectionName, script: string): cockpit.Spawn<string> {
+    return cockpit.script(
+        script,
+        undefined,
+        {
+            err: "message",
+            ...(connectionName === "system" ? { superuser: "try" } : { })
+        });
+}
+
+/* Running virt-xml
+
+   The virtXmlAdd, virtXmlEdit, etc family of functions can be used to
+   run virt-xml with various actions, options, and values. A typical
+   invocation looks like this:
+
+       virtXmlAdd(vm, "graphics", { type: "vnc", port: 5900 });
+
+   This will result in running
+
+       virt-xml <vm-uuid> --add-device --graphics type=vnc,port=5900
+
+   The values are turned into the CSV style of virt-xml in the
+   'obvious' way, according to these rules:
+
+   Properties with "null", "false", or "undefined" as their values are
+   not added to the result at all.  The value "true" is added as the
+   string "yes".
+
+       { type: null, password: "", tls: true }
+       ->
+       password=,tls=yes
+
+   A property name can be the empty string. In that case its value is
+   added with the parent name.
+
+       { source: { "": "abc", mode: "123" } }
+       ->
+       source=abc,source.mode=123
+
+   Arrays are encoded by appending consecutive numbers to the parent.
+
+       { port: [ "a", "b", "c" ] }
+       ->
+       port0=a,port1=b,port2=c
+
+   Top-level values that are neither objects nor arrays are allowed
+   and are encoded without any property name.
+
+       "vnc"
+       ->
+       vnc
+
+   Functions that use a location, namely virtXmlEdit and
+   virtXmlRemove, accept numbers in addition to a match value, just
+   like virt-xml itself.  For example,
+
+       virtXmlEdit(vm, "disk", 1, { bus: "scsi" })
+
+   will run
+
+       virt-xml <vm-uuid> --edit 1 --disk bus=scsi
+
+   Each function also has a "extra_options" argument that can be used
+   to add arbitrary extra options to the virt-xml invocation.  For
+   example,
+
+       virtXmlAdd(vm, "watchdog", "default", { update: true })
+
+   will run
+
+       virt-xml <vm-uuid> --add-device --watchdog default --update
+
+   This is mostly used for hot-plug and non-persistent operations, and
+   there is a shortcut for that: In addition to virtXmlAdd,
+   virtXmlEdit, and virtXmlRemove, there is also a "hot" variant of
+   each, named virtXmlHotAdd, etc.  These function will add the right
+   options to perform a hot-plug when the VM is running, and will also
+   do the right thing with VMs that don't have persistent XML definitions.
+*/
+
+interface virtXmlAction {
+    action: string,
+    location?: undefined | number | unknown,
+    option: string,
+    values: unknown,
+}
+
+async function runVirtXml(
+    vm: VM,
```

Comment:
This only uses connectionName, uuid, and state properties. Given how large `vm` objects are and how often they change, this API forces all components to pass through the whole `vm` and hence re-render a lot more often. Perhaps this can be cleaned up by introducing an object with this small subset of `vm` properties and use that in such APIs?